File: 145-BinaryTreePostorderTraversal/145-BinaryTreePostorderTraversal.py
# Last updated: 4/21/2026, 11:26:44 PM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
   def postorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        self.tr = []
        self._postorder(root)
        return self.tr
   def _postorder(self, root):
        if root is None: return
        self._postorder(root.left)
        self._postorder(root.right)
        self.tr.append(root.val)
        

File: 145-BinaryTreePostorderTraversal/145-BinaryTreePostorderTraversal.cpp
// Last updated: 4/21/2026, 11:26:39 PM
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 #include <vector>
 #include <stack>
 #include <iostream>

class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        stack<TreeNode*> st;
        vector<int> post;
        if (root == NULL) return post;
        TreeNode* curr = root;
        while(curr !=NULL || !st.empty()){
            if(curr != NULL){
                st.push(curr);
                curr = curr->left;
            }
            else{
                TreeNode* temp = st.top()->right;
                if(temp == NULL){
                    temp = st.top();
                    st.pop();
                    post.push_back(temp->val);
                    while(!st.empty() && temp == st.top()->right){
                        temp = st.top();
                        st.pop();
                        post.push_back(temp->val);
                    }
                }
                else{
                    curr = temp;
                }
            }
        }
        return post;
        }
    };
